### PR TITLE
satellite/admin: add coupon creation and listing

### PR DESCRIPTION
**Pull request by stefanbenten**
What: Adds coupon creation and listing.

I had to adjust our coupon handling internally to properly return the information after an insert or update similar to what all other DB interfaces do.

Why: it is useful..

Please describe the tests:
 - Test 1: Adding a coupon
 - Test 2: Listing a coupon

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

**Comment by stefanbenten** [comment]
further onblocking of #3887

File: satellite/admin/user.go [hunk 1]
```diff
@@ -166,3 +175,104 @@ func (server *Server) userInfo(w http.ResponseWriter, r *http.Request) {
...
+
+	var output payments.Coupon
+
+	err = json.Unmarshal(body, &input)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to unmarshal request: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	if input.Duration == 0 {
```
Comment (mniewrzal) [inline satellite/admin/user.go:203]: nit: all ifs cn be switch

File: satellite/admin/user.go [hunk 2]
```diff
@@ -166,3 +175,104 @@ func (server *Server) userInfo(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, "Amount is not set", http.StatusBadRequest)
+		return
+	}
+
+	if input.UserID.IsZero() {
+		http.Error(w, "UserID is not set", http.StatusBadRequest)
+		return
+	}
+
+	// do not change the project limit
```
Comment (mniewrzal) [inline satellite/admin/user.go:218]: is this correct comment?

File: satellite/admin/user.go [hunk 3]
```diff
@@ -166,3 +175,104 @@ func (server *Server) userInfo(w http.ResponseWriter, r *http.Request) {
...
+		Amount:      input.Amount,
+		Duration:    input.Duration,
+		Description: input.Description,
+	})
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to insert coupon: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	output, err = server.db.StripeCoinPayments().Coupons().Get(ctx, coupon.ID)
```
Comment (mniewrzal) [inline satellite/admin/user.go:230]: isn't this just `addCoupon` were you are returning only ID?

File: satellite/admin/README.md [hunk 4]
```diff
@@ -50,10 +50,70 @@ A successful response:
...
+```json
+{
+    "userId":      "12345678-1234-1234-1234-123456789abc",
+    "duration":    2,
+    "amount":      3000,
+    "description": "promotional coupon (valid for 2 billing cycles)"
+}
+```
+
+A successful response (which lists all coupons for the given user):
```
Comment (mniewrzal) [inline satellite/admin/README.md:84]: I think description is wrong.

File: satellite/admin/README.md [hunk 5]
```diff
@@ -50,10 +50,70 @@ A successful response:
...
+
+A successful request:
+
+```json
+{
+    "id": "2fcdbb8f-8d4d-4e6d-b6a7-8aaa1eba4c89"
+}
+```
+
+A successful response (which lists all coupons for the given user):
```
Comment (mniewrzal) [inline satellite/admin/README.md:103]: I think it should something more like `response returns details about specified coupon`.

File: satellite/admin/server.go [hunk 6]
```diff
@@ -34,6 +35,8 @@ type DB interface {
 	ProjectAccounting() accounting.ProjectAccounting
 	// Console returns database for satellite console
 	Console() console.DB
+	// StripeCoinPayments() returns database for satellite stripe coin payments
```
Comment (mniewrzal) [inline satellite/admin/server.go:38]: name shouldn't contain `()`. StripeCoinPayments() -> StripeCoinPayments

File: satellite/admin/server.go [hunk 7]
```diff
@@ -66,6 +69,8 @@ func NewServer(log *zap.Logger, listener net.Listener, db DB, config Config) *Se
 	// When adding new options, also update README.md
 	server.mux.HandleFunc("/api/user", server.addUser).Methods("POST")
 	server.mux.HandleFunc("/api/user/{useremail}", server.userInfo).Methods("GET")
+	server.mux.HandleFunc("/api/user/coupon", server.addCoupon).Methods("POST")
+	server.mux.HandleFunc("/api/user/coupon/{couponid}", server.couponInfo).Methods("GET")
```
Comment (mniewrzal) [inline satellite/admin/server.go:73]: in README its `coupon-id`

File: satellite/admin/user.go [hunk 8]
```diff
@@ -166,3 +174,92 @@ func (server *Server) userInfo(w http.ResponseWriter, r *http.Request) {
...
+
+	w.Header().Set("Content-Type", "application/json")
+	_, _ = w.Write(data) // nothing to do with the error response, probably the client requesting disappeared
+}
+
+func (server *Server) couponInfo(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	vars := mux.Vars(r)
+	id, ok := vars["couponid"]
```
Comment (mniewrzal) [inline satellite/admin/user.go:236]: README is using `coupon-id`

File: satellite/admin/user.go [hunk 9]
```diff
@@ -166,3 +174,92 @@ func (server *Server) userInfo(w http.ResponseWriter, r *http.Request) {
...
+	if !ok {
+		http.Error(w, "couponId missing", http.StatusBadRequest)
```
Comment (mniewrzal) [inline satellite/admin/user.go:238]: depends on final version if variable it should be `coupon-id missing` or `couponid missing`

**Comment by stefanbenten** [comment]
@mniewrzal Our README already has differences for project-id and user-email as well. I kept it consistent in the README using the - separator. In code, we use it without.

**Comment by mniewrzal** [comment]
> @mniewrzal Our README already has differences for project-id and user-email as well. I kept it consistent in the README using the - separator. In code, we use it without.

I see, let's leave it as is for now.

**Comment by kaloyan-raev** [comment]
What if a human mistake is done while creating the coupon? Shouldn't we have a method to delete the coupon, so the admin can start over?

**Comment by stefanbenten** [comment]
@kaloyan-raev I plan on adding that in another PR. I would like to get the tooling in to unblock the covid program asap.

File: satellite/admin/README.md [hunk 10]
```diff
@@ -50,10 +50,69 @@ A successful response:
...
+            "description": "promotional coupon (valid for 2 billing cycles)",
+            "type":        0, 
+            "status":      0,
+            "created":     "2020-05-19T00:34:13.265761+02:00"
+        }
     ]
 }
 ```
 
+## POST /api/user/coupon
```
Comment (kaloyan-raev) [inline satellite/admin/README.md:69]: I think this should be `POST /api/coupon`.

Getting a coupon by ID should be with `GET /api/coupon/{coupon-id}`

This is how we've done it for projects.

File: satellite/admin/user.go [hunk 11]
```diff
@@ -166,3 +174,92 @@ func (server *Server) userInfo(w http.ResponseWriter, r *http.Request) {
...
+	switch {
+	case input.Duration == 0:
+		http.Error(w, "Duration is not set", http.StatusBadRequest)
+		return
+	case input.Amount == 0:
+		http.Error(w, "Amount is not set", http.StatusBadRequest)
+		return
+	case input.UserID.IsZero():
+		http.Error(w, "UserID is not set", http.StatusBadRequest)
+		return
```
Comment (kaloyan-raev) [inline satellite/admin/user.go:208]: The description is displayed in the Deposit & Billing History. It is confusing when there is no description displayed. So perhaps, we should make it mandatory here.

Comment (stefanbenten) [reply satellite/admin/user.go:211]: done

File: satellite/payments/coupons.go [hunk 12]
```diff
@@ -19,7 +19,7 @@ type Coupons interface {
 	ListByUserID(ctx context.Context, userID uuid.UUID) ([]Coupon, error)
 
 	// Create attaches a coupon for payment account.
-	Create(ctx context.Context, coupon Coupon) (err error)
+	Create(ctx context.Context, coupon Coupon) (coup Coupon, err error)
```
Comment (kaloyan-raev) [inline satellite/payments/coupons.go:22]: "coup" could be named as "created"

Comment (stefanbenten) [reply satellite/payments/coupons.go:22]: I'll change that in the PR that adds the remove functionality.

**Pull request merged**